### ISSUE TITLE
Broken links check

### ISSUE THREAD

**Issue by gdiazlo** [wazuh/wazuh-dashboard-plugins#5396]
We must ensure all the links inside the application work correctly before each release.

We need to develop a way to check it automatically. 

All links to our documentation must be versioned to the version used in the dashboard release we're testing.

We must test there are no redirections which might cause a user to end in a different version.

**Comment by Desvelao** [comment]
# Analysis

The Wazuh documentation links are generated using the `webDocumentationLink` function util, that receives a part of the URL, excluding the base path and the version: `${DOCUMENTATION_WEB_BASE_URL}/${version}/${urlPath}`. So the specific URLs are not represented directly in the source code and this could do more difficult the check.

In this scenario, we need some tool or flow that allows to extract the URL sub path, rebuild the complete URL and check if they are online.

Another alternative it could refactor the URL generation to keep a list of the complete URLs or URL sub paths and these could be checked easily but it implies changes in the source code.

## Proposition

There are some tools that can parse the source code and link check:
- [lychee](https://github.com/lycheeverse/lychee) (used by `wazuh-dashboard` repository)

Taking into account the source code contains partial references to URL of the Wazuh documentation, we could extract these URLs if this is not done for another tool and feed to the link checker tool

The extraction of the Wazuh documenation URLs could be done:
> [ts-morph](https://www.npmjs.com/package/ts-morph)
> [semgrep](https://github.com/semgrep/semgrep)
> Custom script

The link check tool can be:
- lychee
- Custom script that do a request (HEAD or GET)

Additionaly, there are other links not related to the Wazuh documenatation that we could check to ensure these are working.

Then a GitHub workflow should be created to check the links.

**Comment by Desvelao** [comment]
# Extract URL sub paths with `ts-morph`

[ts-morph](https://github.com/dsherret/ts-morph) allows to navigate through the AST and could allow to extract the used URL sub paths in the `webDocumentationLink` using a script. The latest versions requires NodeJS 20 or 22 that does not match with the current version of the Wazuh dashboard that uses Node JS 18.

## Spike

I installed the dependency ignoring the engine and created a simple script to extract the first paramenter of the calls into the webDocumentationLink and this can extract the URL sub paths that could allow to rebuild the complete URL and using another tool/script to link check.

[wdp-issue-5396-link-check-extract-url-sub-path-ts-morph.txt](https://github.com/user-attachments/files/21157018/wdp-issue-5396-link-check-extract-url-sub-path-ts-morph.txt)

<details>

<summary>extracting the sub path of the `wazuh` plugin</summary>
  
```
[
    {
      filename: '/home/node/kbn/plugins/main/public/utils/check-plugin-version.tsx',
      wazuh_docs_sub_path: 'user-manual/elasticsearch/troubleshooting.html'
    },
    {
      filename: '/home/node/kbn/plugins/wazuh-core/server/services/configuration-store.ts',
      wazuh_docs_sub_path: 'user-manual/wazuh-dashboard/settings.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/client.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/agent/agent-management/antiflooding.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/labels.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'pci-dss/policy-monitoring.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/wodle-openscap.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/wodle-ciscat.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/wodle-osquery.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/wodle-syscollector.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/active-response.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/wodle-command.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/wodle-docker.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/capabilities/log-data-collection/index.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/server/lib/reporting/agent-configuration.ts',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/syscheck.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/agents/prompts/prompt-agent-never-connected.tsx',
      wazuh_docs_sub_path: 'user-manual/agent/agent-management/agent-connection.html#checking-connection-with-the-wazuh-manager'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/search-bar/query-language/aql.tsx',
      wazuh_docs_sub_path: 'user-manual/api/queries.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/search-bar/query-language/wql.tsx',
      wazuh_docs_sub_path: 'user-manual/wazuh-dashboard/queries.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/tools/devtools/devtools-old.tsx',
      wazuh_docs_sub_path: 'user-manual/api/reference.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/endpoints-summary/dashboard/components/outdated-agents-card.tsx',
      wazuh_docs_sub_path: 'upgrade-guide/wazuh-agent/index.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/security/roles-mapping/components/rule-editor.tsx',
      wazuh_docs_sub_path: 'user-manual/api/rbac/auth-context.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/endpoints-summary/register-agent/components/group-input/group-input.tsx',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/client.html#groups'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/endpoints-summary/register-agent/components/optionals-inputs/optionals-inputs.tsx',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/client.html#enrollment-agent-name'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/endpoints-summary/register-agent/components/server-address/server-address.tsx',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/client.html#manager-address'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/endpoints-summary/register-agent/containers/steps/steps.tsx',
      wazuh_docs_sub_path: 'user-manual/agent/agent-enrollment/security-options/using-password-authentication.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/overview/fim/common/hocs/validate-fim-states-index-pattern.tsx',
      wazuh_docs_sub_path: 'user-manual/capabilities/file-integrity/index.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/overview/it-hygiene/common/hocs/validate-system-inventory-index-pattern.tsx',
      wazuh_docs_sub_path: 'user-manual/capabilities/system-inventory/index.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/overview/vulnerabilities/common/components/check-module-enabled.tsx',
      wazuh_docs_sub_path: 'user-manual/capabilities/vulnerability-detection/configuring-scans.html#configuring-vulnerability-detection'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/overview/vulnerabilities/common/hocs/validate-vulnerabilities-states-index-pattern.tsx',
      wazuh_docs_sub_path: 'user-manual/capabilities/vulnerability-detection/index.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/components/endpoints-summary/register-agent/components/os-selector/os-card/os-card.tsx',
      wazuh_docs_sub_path: 'installation-guide/wazuh-agent/index.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/controllers/management/components/management/configuration/github/github.tsx',
      wazuh_docs_sub_path: 'cloud-security/github/index.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/controllers/management/components/management/configuration/github/github.tsx',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/github-module.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/controllers/management/components/management/configuration/office365/constants.tsx',
      wazuh_docs_sub_path: 'cloud-security/office365/index.html'
    },
    {
      filename: '/home/node/kbn/plugins/main/public/controllers/management/components/management/configuration/office365/constants.tsx',
      wazuh_docs_sub_path: 'user-manual/reference/ossec-conf/office365-module.html'
    }
  ]
```
</details>

**Comment by Desvelao** [comment]
# Extract URL sub paths with `semgrep`

[semgrep](https://github.com/semgrep/semgrep) is a static analysis tool based on python that can be used to extract the URL subpaths:

## Spike

1. Install the library
> I used 1.89.0 version that provides information about the metavariables (extracted URL sub paths). The latest versions could require an account or I could not get the output json file had the `metavars` property.
```
pip install semgrep==1.89.0
```

2. Define the configuration for semgrep (create a `semgrep.conf.yml` file):
```
rules:
  - id: find-url-sub-path-webdocumentationlink-function
    languages: [javascript, typescript]
    message: 'Find URL sub path used in webDocumentationLink'
    severity: WARNING
    pattern: webDocumentationLink('$SUBPATH')
    metadata:
      category: maintainability
```

3. Scan the source code and output to a JSON file
```
semgrep scan --config=/path/to/semgrep.conf.yml --exclude node_modules/ --json --output=subpaths.json <SRC>
```
where:
- `<SRC>`: path to source code

4. Get the URL sub paths:

```
jq -r '.results[].extra.metavars["$SUBPATH"].abstract_content' subpaths.json
```

<details>

This is a output example:

```
user-manual/agent/agent-enrollment/index.html

user-manual/agent/agent-management/agent-connection.html#checking-connection-with-the-wazuh-manager
upgrade-guide/wazuh-agent/index.html
upgrade-guide/wazuh-agent/index.html
installation-guide/wazuh-agent/index.html
user-manual/agent/agent-enrollment/security-options/using-password-authentication.html
user-manual/wazuh-server-cluster/cluster-nodes-configuration.html
user-manual/capabilities/file-integrity/index.html
user-manual/capabilities/system-inventory/index.html
user-manual/capabilities/vulnerability-detection/configuring-scans.html#configuring-vulnerability-detection
user-manual/capabilities/vulnerability-detection/index.html
user-manual/api/queries.html
user-manual/wazuh-dashboard/queries.html
user-manual/api/rbac/auth-context.html
user-manual/api/reference.html
user-manual/capabilities/active-response/index.html
user-manual/reference/ossec-conf/active-response.html
user-manual/capabilities/active-response/index.html
user-manual/reference/ossec-conf/active-response.html
user-manual/capabilities/active-response/index.html
user-manual/reference/ossec-conf/commands.html
user-manual/capabilities/agentless-monitoring/index.html
user-manual/reference/ossec-conf/agentless.html
user-manual/manager/alert-management.html#configuring-email-alerts
user-manual/manager/alert-management.html#smtp-server-with-authentication
user-manual/reference/ossec-conf/email-alerts.html
getting-started/use-cases/index.html
user-manual/reference/ossec-conf/alerts.html
user-manual/agent/agent-management/labels.html
user-manual/reference/ossec-conf/labels.html
user-manual/reference/ossec-conf/reports.html
user-manual/manager/alert-management.html#configuring-syslog-output
user-manual/reference/ossec-conf/syslog-output.html
cloud-security/amazon/index.html
user-manual/reference/ossec-conf/wodle-s3.html
cloud-security/azure/index.html
user-manual/reference/ossec-conf/wodle-azure-logs.html
user-manual/agent/agent-management/antiflooding.html
user-manual/reference/ossec-conf/client-buffer.html
user-manual/agent/agent-management/agent-connection.html#checking-connection-with-the-wazuh-manager
user-manual/reference/ossec-conf/client.html
user-manual/wazuh-server-cluster/cluster-nodes-configuration.html
user-manual/reference/ossec-conf/cluster.html
user-manual/reference/ossec-conf/wodle-command.html
user-manual/manager/index.html
user-manual/capabilities/index.html
user-manual/reference/ossec-conf/index.html
user-manual/capabilities/container-security/use-cases.html
user-manual/reference/ossec-conf/wodle-docker.html
cloud-security/github/index.html
user-manual/reference/ossec-conf/github-module.html
user-manual/reference/ossec-conf/global.html
user-manual/reference/ossec-conf/logging.html
user-manual/reference/daemons/wazuh-remoted.html
user-manual/reference/ossec-conf/remote.html
cloud-security/gcp/index.html
user-manual/reference/ossec-conf/gcp-pubsub.html
user-manual/manager/integration-with-external-apis.html
user-manual/capabilities/malware-detection/virus-total-integration.html
user-manual/reference/ossec-conf/integration.html
user-manual/capabilities/file-integrity/index.html
user-manual/reference/ossec-conf/syscheck.html
user-manual/capabilities/system-inventory/index.html
user-manual/reference/ossec-conf/wodle-syscollector.html
user-manual/capabilities/log-data-collection/index.html
user-manual/reference/ossec-conf/localfile.html
user-manual/capabilities/log-data-collection/monitoring-log-files.html
user-manual/reference/ossec-conf/socket.html
cloud-security/office365/index.html
user-manual/reference/ossec-conf/office365-module.html
user-manual/capabilities/system-inventory/osquery.html
user-manual/reference/ossec-conf/wodle-osquery.html
user-manual/capabilities/malware-detection/index.html
user-manual/reference/ossec-conf/rootcheck.html
user-manual/agent/agent-enrollment/index.html
user-manual/reference/ossec-conf/auth.html
user-manual/capabilities/vulnerability-detection/index.html
user-manual/reference/ossec-conf/vuln-detector.html
user-manual/wazuh-dashboard/troubleshooting.html#saved-object-for-index-pattern-not-found
user-manual/reference/ossec-conf/client.html
user-manual/agent/agent-management/antiflooding.html
user-manual/reference/ossec-conf/labels.html
user-manual/reference/ossec-conf/wodle-osquery.html
user-manual/reference/ossec-conf/wodle-syscollector.html
user-manual/reference/ossec-conf/active-response.html
user-manual/reference/ossec-conf/wodle-command.html
user-manual/reference/ossec-conf/wodle-docker.html
user-manual/capabilities/log-data-collection/index.html
user-manual/reference/ossec-conf/syscheck.html
user-manual/agent/agent-enrollment/index.html

user-manual/wazuh-dashboard/settings.html
user-manual/agent/agent-enrollment/index.html
user-manual/agent/agent-enrollment/index2.html
```
</details>

**Comment by Desvelao** [comment]
# Link check with `lychee`

[lychee](https://github.com/lycheeverse/lychee) can be used to check if the links are available.

If we keep the source code that generates the links to the Wazuh documentation using the webDocumentationLink function, we need a way to extract these URLs to be able to check them using lychee.

Wazuh dashboard uses lychee to check links:
- GitHub action: https://github.com/wazuh/wazuh-dashboard/blob/v4.12.0/.github/workflows/links_checker.yml#L21
- `.lycheeignore`: https://github.com/wazuh/wazuh-dashboard/blob/v4.12.0/.lycheeignore

**Comment by Desvelao** [comment]
# Extracting the URL sub paths with Linux tools

I was exploring to extrating the URL sub path used in the `webDocumentationLink` using some linux tools.

## Spike

The following script can extract the URL sub paths and rebuild the adding the prefix with the base URL and Wazuh version (manjor.minor) sending the URLs to a file that could be used with lychee or other tool to link check:

I created a script that extract the URLs getting from the `webDocumentationLink` calls (excluding the test files):

<details>
  <summary>extract_web_doc_links.sh</summary>

```bash
#!/bin/bash

# List of required tools
tools=("find" "grep" "jq" "sed" "perl")

# Check each tool
for tool in "${tools[@]}"; do
    if ! command -v "$tool" &> /dev/null; then
        echo "❌ Error: $tool is not installed."
        missing=true
    fi
done


# Optional exit if any tool is missing
if [ "$missing" = true ]; then
    echo "🚨 Please install the missing tools before proceeding."
    exit 1
fi

log(){
  echo $@ >&2
}

DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
PARENT_DIR="$(dirname "$DIR")"
ROOT_PATH="$(dirname "$PARENT_DIR")"

DESTINATION_LINKS_FILE='wz_docs_links.txt'

WEB_DOCS_BASE_URL=$(grep 'DOCUMENTATION_WEB_BASE_URL' plugins/main/common/constants.ts | sed -E "s/.*= '(.*)';/\1/")
WZ_VERSION=$(jq -r '.version | split(".") | .[0] + "." + .[1]' plugins/main/package.json)
WEB_DOCS_PREFIX_CURRENT="$WEB_DOCS_BASE_URL/$WZ_VERSION/"

log 'Parsing webDocumentationLink calls with 1 parameter excluding the test files'
find $ROOT_PATH -type f \( -name '*.js' -o -name '*.ts' -o -name '*.jsx' -o -name '*.tsx' \) \
  ! -name '*.test.js' ! -name '*.test.jsx' ! -name '*.test.ts' ! -name '*.test.tsx' \
  -not -path '*/node_modules/*' \
  -exec grep -Pzoh "webDocumentationLink\\s*\\(\\s*['\"]\\K[\\s\\S]*?(?=['\"][ \\t]*[,\\)])" {} + \
  | tr '\0' '\n' \
  | sort -u \
  | sed '/^$/d' \
  |  sed "s|^|$WEB_DOCS_PREFIX_CURRENT|"

log 'Parsing webDocumentationLink calls with 2 parameters excluding the test files'
find $ROOT_PATH -type f \( -name '*.js' -o -name '*.ts' -o -name '*.jsx' -o -name '*.tsx' \) \
  ! -name '*.test.js' ! -name '*.test.jsx' ! -name '*.test.ts' ! -name '*.test.tsx' \
  -not -path '*/node_modules/*' -exec perl -0777 -nle '
    while (
      /webDocumentationLink\(\s*["\047]([^"\047]+)["\047]\s*,\s*["\047]([^"\047]+)["\047]/gs
    ) {
      print "$2/$1\n"
    }
  ' {} + \
  | sort -u \
  | sed '/^$/d' \
  | sed "s|^|$WEB_DOCS_BASE_URL/|"

```
</details>

Run the tool

```console
bash extract_web_doc_links.sh > wz_docs_links.txt
```

Then, we could use the generated file with the links to feed to some link checker tool such as lychee:

<details>

  <summary>link check with lychee</summary>

```
lychee -v wz_docs_links.txt 
     [200] https://documentation.wazuh.com/4.8/user-manual/capabilities/policy-monitoring/openscap/index.html
     [200] https://documentation.wazuh.com/4.4/user-manual/elasticsearch/troubleshooting.html
     [200] https://documentation.wazuh.com/4.8/user-manual/capabilities/policy-monitoring/ciscat/ciscat.html
     [200] https://documentation.wazuh.com/4.8/user-manual/capabilities/policy-monitoring/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/manager/alert-management.html#configuring-syslog-output
     [200] https://documentation.wazuh.com/4.13/installation-guide/wazuh-agent/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/wazuh-dashboard/queries.html
     [200] https://documentation.wazuh.com/4.13/user-manual/manager/integration-with-external-apis.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/auth.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/wodle-osquery.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/wodle-s3.html
     [200] https://documentation.wazuh.com/4.13/user-manual/api/rbac/auth-context.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/client.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/office365-module.html
     [200] https://documentation.wazuh.com/4.13/user-manual/wazuh-dashboard/settings.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/vulnerability-detection/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/vulnerability-detection/configuring-scans.html#configuring-vulnerability-detection
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/client-buffer.html
     [200] https://documentation.wazuh.com/4.13/user-manual/agent/agent-enrollment/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/agentless-monitoring/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/api/reference.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/wodle-openscap.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/log-data-collection/monitoring-log-files.html
     [200] https://documentation.wazuh.com/4.13/user-manual/api/queries.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/malware-detection/virus-total-integration.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/global.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/malware-detection/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/elasticsearch/troubleshooting.html
     [200] https://documentation.wazuh.com/4.13/user-manual/agent/agent-management/antiflooding.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/localfile.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/policy-monitoring/openscap/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/policy-monitoring/ciscat/ciscat.html
     [200] https://documentation.wazuh.com/4.13/cloud-security/gcp/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/manager/automatic-reports.html
     [200] https://documentation.wazuh.com/4.13/cloud-security/amazon/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/commands.html
     [200] https://documentation.wazuh.com/4.13/upgrade-guide/wazuh-agent/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/active-response.html
     [200] https://documentation.wazuh.com/4.13/cloud-security/github/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/wazuh-dashboard/troubleshooting.html#saved-object-for-index-pattern-not-found
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/agentless.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/alerts.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/cluster.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/socket.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/wodle-command.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/rootcheck.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/github-module.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/remote.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/client.html#enrollment-agent-name
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/wodle-syscollector.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/labels.html
     [200] https://documentation.wazuh.com/4.13/user-manual/manager/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/policy-monitoring/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/agent/agent-enrollment/security-options/using-password-authentication.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/file-integrity/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/system-inventory/osquery.html
     [200] https://documentation.wazuh.com/4.13/pci-dss/policy-monitoring.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/client.html#groups
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/logging.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/container-security/use-cases.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/active-response/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/email-alerts.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/system-inventory/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/wodle-ciscat.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/client.html#manager-address
     [200] https://documentation.wazuh.com/4.2/pci-dss/policy-monitoring.html
     [200] https://documentation.wazuh.com/4.8/user-manual/reference/ossec-conf/wodle-openscap.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/daemons/wazuh-remoted.html
     [200] https://documentation.wazuh.com/4.13/user-manual/agent/agent-management/agent-connection.html#checking-connection-with-the-wazuh-manager
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/integration.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/gcp-pubsub.html
     [200] https://documentation.wazuh.com/4.13/user-manual/agent/agent-management/labels.html
     [200] https://documentation.wazuh.com/4.13/cloud-security/azure/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/reports.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/syscheck.html
     [200] https://documentation.wazuh.com/4.13/user-manual/manager/alert-management.html#configuring-email-alerts
     [200] https://documentation.wazuh.com/4.13/getting-started/use-cases/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/wodle-azure-logs.html
     [200] https://documentation.wazuh.com/4.13/user-manual/wazuh-server-cluster/cluster-nodes-configuration.html
     [200] https://documentation.wazuh.com/4.13/user-manual/capabilities/log-data-collection/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/manager/alert-management.html#smtp-server-with-authentication
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/wodle-docker.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/vuln-detector.html
     [200] https://documentation.wazuh.com/4.13/cloud-security/office365/index.html
     [200] https://documentation.wazuh.com/4.13/user-manual/reference/ossec-conf/syslog-output.html
     [200] https://documentation.wazuh.com/4.8/user-manual/reference/ossec-conf/wodle-ciscat.html
     [200] https://documentation.wazuh.com/4.6/user-manual/manager/automatic-reports.html

```
</details>

**Comment by Desvelao** [comment]
I moved to backlog until a target release is assigned.

**Comment by guidomodarelli** [comment]
## 🧩 **Automatic link validation in documentation**

### 1️⃣ **Main Objective**

Automate the validation of links used, ensuring they remain active and valid, especially those that may have been removed in previous versions (such as 4.12, 4.13, 4.14, etc.).

---

### 2️⃣ **Technical Context of the Problem**

* The `webDocumentationLink()` function is used in the source code to generate links to the documentation.
* This function receives a **subpath** as an argument and constructs the final URL by adding a **base prefix** and, optionally, the **Wazuh version**.
* The issue: the full URL is not included in the source code, which complicates automatic validation with external tools.

---

### 3️⃣ **Proposed Solutions and Alternatives**

#### 🔹 Option A: Parse from code without modifying the structure

* Detect calls to `webDocumentationLink()` to extract the subpaths.
* Reconstruct the URLs by adding:

  * Wazuh base prefix.
  * Corresponding version (if applicable).

#### 🔹 Option B: Define paths in an external file

* Declare the full paths (or at least the subpaths) in a centralized file.
* Facilitate reading by validation tools.

---

### 4️⃣ **Tools Evaluated**

| Tool              | Status / Comment                                                                 |
| ----------------- | -------------------------------------------------------------------------------- |
| **Lychee**        | Parses source code but cannot work with incomplete URLs without an extra plugin. |
| **ts-morph**      | Works, but requires Node 20+ (currently using Node 18).                          |
| **semgrep**       | Tested. Requires extra configuration.                                            |
| **Custom Script** | Viable alternative. Would make GET requests to validate status (HTTP 200).       |

---

### 5️⃣ **Limitations Found**

* **Node 20 required** for some tools (current version: Node 18).
* Dependency on utilities like `jq`, which may not be installed on all machines.
* Complexity of using **regular expressions** in Bash.
* Python implementation is avoided due to portability and stack consistency concerns.

---

### 6️⃣ **Tests Performed**

* Antonio created a **Bash script** that:

  * Extracts subpaths from `webDocumentationLink()` calls.
  * Reconstructs complete URLs.
  * Validates if they are active via HTTP requests.

* Tests were performed with cases such as:

  * Simple subpaths.
  * Subpaths with specified version (`4.8`, for example).
  * Validation via workflows (GitHub Actions).

---

### 7️⃣ **Execution Proposal**

1️⃣ Develop a **portable** solution (ideally in Bash) that:

* Does not require many external dependencies.
* Works on multiple Linux distros.

2️⃣ Alternatively, encapsulate in a **Dockerfile** to bypass environment limitations (Node version, dependencies, etc.).

3️⃣ Create a **workflow in GitHub Actions** that:

* Runs on pull requests.
* Automatically validates generated links.

---

### 8️⃣ **Final Recommendations**

* Prioritize a simple, portable, and extensible implementation.
* Consider tools like ts-morph within a Docker container to avoid version conflicts.
* Avoid Python unless strictly necessary.

---

### 🧠 Additional Notes

* Few special cases were found with older versions (like 4.2, 4.8).

**Comment by guidomodarelli** [comment]
### Observations

There are four cases where the `webDocumentationLink` function is used with a first parameter that is a constant imported from the `constants` file. Since this value is an external constant and not a string literal, it is not possible to statically determine whether the resulting URL is valid.

However, in theory, static validation could be performed if the `constants` file is analyzed and the value of the constant used is resolved, as long as it is statically determinable. In that case, the constant could be replaced by its value and the resulting URL could be checked for validity.

Files involved:

* [`header.tsx`](https://github.com/wazuh/wazuh-dashboard-plugins/blob/c25fef37fe07dea8aed80f9a3b9af9d8a9a11dc2/plugins/main/public/components/settings/configuration/components/header.tsx#L65-L67)
* [`check-api.service.ts`](https://github.com/wazuh/wazuh-dashboard-plugins/blob/c25fef37fe07dea8aed80f9a3b9af9d8a9a11dc2/plugins/main/public/components/health-check/services/check-api.service.ts#L57-L59)
* [`check-setup.service.ts`](https://github.com/wazuh/wazuh-dashboard-plugins/blob/c25fef37fe07dea8aed80f9a3b9af9d8a9a11dc2/plugins/main/public/components/health-check/services/check-setup.service.ts#L58-L60)
* [`wz-blank-screen.js`](https://github.com/wazuh/wazuh-dashboard-plugins/blob/c25fef37fe07dea8aed80f9a3b9af9d8a9a11dc2/plugins/main/public/components/wz-blank-screen/wz-blank-screen.js#L103-L105)

In addition, there is one more case where a dynamic variable is used as a parameter. In that scenario, static validation is indeed impossible, since the value cannot be known at analysis time.

* [`href={webDocumentationLink(documentationLink)}`](https://github.com/wazuh/wazuh-dashboard-plugins/blob/c25fef37fe07dea8aed80f9a3b9af9d8a9a11dc2/plugins/main/public/components/settings/configuration/components/categories/components/category/category.tsx#L84)

One possible solution to address these cases—especially the last one involving a dynamic value—is to change how the `webDocumentationLink` function is used. Instead of calling it inside the `Category` component with a potentially dynamic value, we could pass the fully constructed documentation link directly as a prop to the component.

This way, the `webDocumentationLink` function would only be used in places where the specific part of the URL is statically known. In other words, rather than centralizing its usage inside the component, we would use it at the call sites where the documentation link is actually determined.

This approach would allow for static validation of the links, since the values would be explicitly defined in each usage context, avoiding reliance on dynamic variables that are difficult to analyze statically.

**Comment by guidomodarelli** [comment]
# Update 2025/07/15

* 🧪 **Static analysis with `ts-morph`:**

  * I implemented an initial solution based on `ts-morph` to validate the links generated by `WebDocumentationLink`.
  * I found that the library does not allow resolving values when the first parameter is a variable, which prevents proper validation of URLs in those cases.

* 🔄 **Change in link validation approach:**

  * We proposed centralizing the links in an external file per plugin (one for core and another for main).
  * We decided to compile those files with TypeScript and then analyze them with `lychee` to validate the links on the compiled code.

* ⚙️ **Automation in CI:**

  * We planned to automatically run the link validation process on each pull request through a workflow in GitHub Actions.

**Comment by guidomodarelli** [comment]
### 🔍 Spike: Validation of the Fragment Identifier in URLs (`#fragment`)

As part of the research on how to programmatically validate the final part of a URL containing a `fragment identifier` (e.g., `#enrollment-agent-name`), the behavior of different HTTP request methods and their relationship with the fragment was analyzed.

#### ✅ Key Observations

* **The fragment identifier (`#...`) is not sent to the server** in any standard HTTP request (`GET`, `HEAD`, `POST`, etc.). This applies to tools like `curl` as well as browsers.
* **The fragment is processed exclusively by the client** (typically the browser), which uses it to scroll to the HTML document element with that `id`.
* Therefore, tools like `curl` will return a `200 OK` for the base URL (`client.html`), regardless of whether the fragment exists in the DOM.

#### 🛠️ Validation Alternatives

Since the fragment is not part of the request, the returned HTML content must be inspected to verify its existence. Two approaches were explored:

1. **Manual verification via console**:

   ```bash
   curl -s https://dominio.com/user-manual/reference/ossec-conf/client.html | grep 'id="enrollment-agent-name"'
   ```

2. **Automated validation via Node.js script (`axios` + `cheerio`)**:

   ```javascript
   const axios = require('axios');
   const cheerio = require('cheerio');

   const url = 'https://dominio.com/user-manual/reference/ossec-conf/client.html';
   const fragment = 'enrollment-agent-name';

   axios.get(url)
     .then(response => {
       const $ = cheerio.load(response.data);
       const exists = $(`#${fragment}`).length > 0;
       console.log(exists ? 'Fragment exists' : 'Fragment not found');
     })
     .catch(err => {
       console.error('Error fetching URL:', err.message);
     });
   ```

   To run this script, make sure the dependencies are installed:

   ```bash
   npm install axios cheerio
   ```

#### 📌 Conclusion

It is not possible to directly verify on the server side whether a fragment is valid using `curl` or other traditional networking tools. To validate the existence of a fragment identifier, it is necessary to inspect the HTML content of the base document and check if a node with the corresponding `id` attribute exists.

**Comment by guidomodarelli** [comment]
# Update 2025/07/23

I tried to apply these changes to be able to check all the URLs found throughout the repository and I couldn't get it to work; for some reason, the URLs in the links.txt file are valid when I access them by clicking manually, but the tool in charge of verifying that the URL is valid reports the URL as an error and I don't know why.

```diff
diff --git a/docker/url-checks/Dockerfile b/docker/url-checks/Dockerfile
index 81e8b655d..ab7e2d08e 100644
--- a/docker/url-checks/Dockerfile
+++ b/docker/url-checks/Dockerfile
@@ -10,6 +10,8 @@ RUN npm install -g tsx typescript --no-package-lock
 # Copy the entire repository
 COPY . .
 
+RUN apk add --no-cache ripgrep && rg -o 'https?://[^\s"'"'"'<>()]+' plugins/ --glob '!package-lock.json' --glob '!yarn.lock' --glob '!*.test.*' | grep -vE '\$\{' > links.txt
+
 # Extract URLs
 RUN tsx scripts/url-checks/extractor.ts
 
@@ -19,7 +21,8 @@ FROM lycheeverse/lychee:latest AS url-checker
 WORKDIR /app
 
 # Copy extracted URLs from previous stage
-COPY --from=url-extractor /app/scripts/url-checks/extracted-urls.md scripts/url-checks/extracted-urls.md
+COPY --from=url-extractor /app/scripts/url-checks/extracted-urls.md /app/extracted-urls.md
+COPY --from=url-extractor /app/links.txt /app/links.txt
 
 # Default command to check URLs
-CMD ["--max-redirects", "0", "--format", "detailed", "--no-progress", "scripts/url-checks/extracted-urls.md"]
+CMD ["--max-redirects", "0", "--format", "detailed", "--no-progress", "extracted-urls.md", "links.txt"]
```

**Comment by guidomodarelli** [comment]
With this command:

```bash
curl -s https://api.github.com/repos/wazuh/wazuh-dashboard/releases/latest \
  | jq '.tag_name' -r \
  | grep -Po '\d+\.\d+\.\d+'
```

it’s possible to get the **latest stable version of Wazuh Dashboard** released in production.
This allows us to build and verify URLs always based on the most recent version available.

It’s important to keep in mind that this method could fail in one specific scenario:
if a new version adds documentation that didn’t exist in previous versions, those URLs won’t be available in the latest version and will therefore return an error. However, this isn’t really a problem, since in that case we know the page doesn’t exist in the latest stable version and will only be available starting from a future release.

In summary, this technique lets us check the **vast majority of URLs** —and in most cases, all of them— with the only exception being content that was added recently and is not yet included in the current production version.

**Comment by guidomodarelli** [comment]
# Update 2025/08/26

## Summary of the discussion on “Broken links check”

**Current status**

* The PR uses **lychee** to validate URLs against the **latest published release** (e.g., 4.12.0).
* Reason: if we compare against the version in `package.json` (in development, e.g., 4.14.0), the docs for that version **don’t exist yet**, and the checks fail on every PR.

**Problem**

* As it stands, we could **ship a release with broken links** added or modified in the new version, since they’re only detected **after** publishing the release.

**Improvement options considered**

1. **Validate against the “pre-release” docs** (flag `--pre-release`):

   * In PRs, point links to the **pre-release** version of the documentation.
   * Advantage: detects breakages **before** publishing.
   * Requires: mode parameter in the script and pre-release doc source/URL.

2. **Validate against the target version’s documentation repo**:

   * Spin up the **documentation** repo locally (or in CI) for the target version and run lychee against that instance (e.g., `http://localhost:PORT/...`).
   * Advantage: maximum fidelity with the actual content that will ship.
   * Requires: logic changes (resolve base URL per version, extra CI step).

**Draft implementation proposal**

* New script flag:

  * `--mode=release` (current default): validates against **latest published release**.
  * `--mode=pre-release`: validates against **pre-release docs**.
  * `--mode=local-docs`: points to a **local URL** to validate the documentation repo spun up in CI.

**Conclusion**

* The current check avoids false positives in PRs but **does not prevent** broken links in the new release.
* With `--pre-release` and/or validation against the target documentation repo, we **do prevent** broken links from reaching production.
